### ISSUE TITLE
Downloads inside HTML5 apps do not work

### ISSUE THREAD

**Issue by jredrejo** [learningequality/kolibri#11987]
## Observed behavior

Some HTML5 apps include content that can be donwloaded, as documentation, images, etc.
Current iframe sandbox in Kolibri blocks them.

## Errors and logs
Just clicking on a download link does not produce any effect

There is an  [allow-downloads](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#allow-downloads) attribute for iframe that might be enabled globally in Kolibri to allow this.

## Expected behavior
Clicking in a link to download a resource should work inside HTML5 apps

As this introduces security issues in some cases ( it would allow HTML5 apps to initiate downloads programmatically as well as through user interaction), the ideal solution would be adding an option to the `HTML5AppNode` to use it (being `false` by default)

## User-facing consequences

Users can not use all the contents the app includes

## Steps to reproduce

For example, courses or tutorial including documentation for further reading 
![image](https://github.com/learningequality/kolibri/assets/1008178/f9377377-8a26-4594-884f-7341f2e55269)
don't allow this documentation to be downloaded, even if the document is included in the `HTML5AppNode`

## Context
kolibri 0.16.0

**Comment by poju3185** [comment]
Hi @jredrejo, could you give me an example of such html5 app?

**Comment by jredrejo** [comment]
> Hi @jredrejo, could you give me an example of such html5 app?



Yes, @poju3185  here's one: https://articulateusercontent.com/review/uploads/cPj4SORrydR49zuKNHfjfUrAe2RyfIyd/DxkvFL4f/index.html#/lessons/HCQgFWFRBDiv8VRaUq_uEDqzaF-57tXg

Anyway, it's important to understand the broader implications of enabling downloads through iframes. This goes beyond a simple code change.

Implementing this change would require:

- Modifying https://github.com/learningequality/studio to handle apps with and without downloadable content.
- Similar changes to https://github.com/learningequality/ricecooker to allow channel creation with the download property enabled or disabled.
- Ensuring Kolibri can manage this functionality only for apps that explicitly allow downloads. As noted in https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#allow-downloads, the "allow-downloads" attribute is global but has security limitations.

Finally, the most crucial aspect is mitigating the risk of malicious content. How can we ensure uploaded content doesn't cause harm, and how much can we trust imported content? Cryptography might be a solution, but allowing anyone to upload apps with download capabilities opens up security vulnerabilities.

As you can see, this is a complex feature with significant technical and architectural considerations. Its feasibility needs careful evaluation.

**Comment by poju3185** [comment]
Thank you for the detailed explanation! Given the complexity and the careful evaluation needed to ensure both functionality and security, I think I'll sit this one out and observe for now. I look forward to seeing how this develops and learning from the process.
P.S. I am able to download the attachment file in the provided link. Is this the expected behavior?

**Comment by jredrejo** [comment]
> P.S. I am able to download the attachment file in the provided link. Is this the expected behavior?

yes, but  if this html app is added in a channel for kolibri you can't , that's what this issue tries to change

**Comment by poju3185** [comment]
I see. Thank you.

**Comment by Wck-iipi** [comment]
@jredrejo I see that this issue has the tag TODO:needs decision. But you have specified in great detail on what to do. If this issue is up for grabs, I would like to solve this issue.

**Comment by rtibbles** [comment]
There's still a strong concern here about the potential for malicious behaviour. Loosening the sandbox, even selectively, would allow programmatically initiated downloads as well as user initiated ones.

I am not sure that allowing any loosening of the sandbox is actually desirable, but rather allowing non-renderable files to be included as separate file attachments, which the user can then choose to download via the user interface.

Either way, this issue is not ready for contribution, hence the needs decisions label.

**Comment by jaltekruse** [comment]
I wanted to chime in to share some instances where I have run into this limitation of Kolibri. I am currently trying to work on getting the MIT Scratch language/ game development platform added as a channel to Kolibri.  I made a post in the forums about my progress so far, you can import it into an instance of Kolibri with the code: jitig-safid -  https://community.learningequality.org/t/scratch-block-based-programming-channel-work-in-progress/3905

The biggest issue I know of with the app as it stands right now is that projects cannot be saved, seemingly because of this limitation of the sandbox. The scratch platform hosted online has a SaaS based cloud storage for projects, but as a fallback it allows saving the projects to a device as files. This would be an instance of a problematically initiated download, although I have some vague idea that browsers have some way of tracking if some javascript was initiated by a user interaction like a click and attempt to prevent bad behavior being initiated from simply visiting a site, I have gotten permissions popups about "this site is trying to download multiple files" that I believe are related to this.

I am planning to work on a modified version of Scratch that will save the projects inside the browser in IndexedDB which I know is available in the sandbox. And while I think this is worthwhile to pursue as a first version, it does have limitations, namely that student work will be stuck tied to a single device and a single point of failure that is hard to back up (I do understand that Kolibri installs aren't often fancy setups with mirrored databases to offsite backup and such, but there are features included to enable backup of student data that is saved into the platform).

I ran into a similar issue bringing my own FOSS software for managing digital math homework, which currently has a file-based workflow and experiences the same behavior of being allowed to open files from the local device but not save them when I bundle the offline webapp of that site into a kolibri channel.

I would advocate that giving users the agency to manage files themselves for use cases like this could warrant enabling this feature in the sandbox, and that moderation of the content library to keep out malicious software might be a messy but ultimately necessary part of enabling rich experiences in the Kolibri ecosystem.
